mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -1429,60 +1535,67 @@ createCPUConvFunc(ModuleOp module,
   return func;
 }
 
-std::string getTypeStr(const mlir::Type &type) {
-  std::string typeName;
-  llvm::raw_string_ostream nameStream(typeName);
-  nameStream << type;
-  return nameStream.str();
-}
-
 static func::FuncOp getMemcpyFuncDecl(ModuleOp &module,
-                                      const mlir::Type &srcElemType,
-                                      const mlir::Type &dstElemType) {
+                                      const mlir::MemRefType srcType,
+                                      const mlir::MemRefType destType) {
   OpBuilder b(module.getContext());
 
-  // memcpy_<srcElemType>_<dstElemType>
+  assert(srcType.getRank() == 1 && "Memcopy takes 1-D sources");
+  assert(destType.getRank() == 1 && "Memcopy takes 1-D destinations");
+
+  mlir::Type srcElemType = srcType.getElementType();
+  mlir::Type dstElemType = destType.getElementType();
+  // memcpy_<srcElemType>_<dstElemType>_(srcSize|any)
   std::string funcName = "_memcpy_";
-  funcName += getTypeStr(srcElemType);
-  funcName += "_";
-  funcName += getTypeStr(dstElemType);
+  llvm::raw_string_ostream funcNameStr(funcName);
+  funcNameStr << srcElemType << "_" << dstElemType << "_";
+
+  int64_t numElements = -1;
+  if (srcType.hasStaticShape())
+    numElements = srcType.getNumElements();
+
+  if (numElements != -1)
+    funcNameStr << numElements;
+  else
+    funcNameStr << "any";
+
+  if ((numElements == -1 && !destType.hasStaticShape()) ||
+      numElements != destType.getNumElements())
```

Comment:
Does `getNumElements()` returns -1 if the type does not have static shape?

Reply:
Ok, so, my best guess - I might have to try this - is that if you pass in a `memref<*xT>` `getNumElements()` will, by way of `getShape()`, explode, but a `memref<?xT>` (for any number of `?`) will also explode by way of the computation going negative, which touches off an assert about integer overflow (or maybe something earlier)